### ISSUE TITLE
Auto-Switching QR-Code Display

### ISSUE THREAD

**Issue by KST1810** [corona-warn-app/cwa-wishlist#413]
As plenty of restaurants, pubs etc have TV's or similar displays, would there be a chance to create an "CWA QR-Code display-app" that automatically iterates through a number of codes to prevent the aforementioned attacks?

_Originally posted by @KST1810 in https://github.com/corona-warn-app/cwa-wishlist/issues/399#issuecomment-808516326_

On the note of creating QR-Codes and iterating through several of them to prevent profiling attacks: 

We could create a simple QR-Code Display app that's optimized for android TV and tablets which is able to iterate through these codes. This enhances the usability and user friendliness for the venues as the managers will no longer need to print out sheets of paper displaying these codes.

**Comment by Ein-Tim** [comment]
Just FYI, the app itself (building it locally) shows this reg. the switching of QR-Codes:

<img src="https://user-images.githubusercontent.com/67682506/112906662-d5e26e00-90ec-11eb-8336-60f17c4b159c.png" width="500">
<img src="https://user-images.githubusercontent.com/67682506/112906872-383b6e80-90ed-11eb-952f-6fcbe143b1ad.png" width="500">

So this display shouldn't change the code every 1 hour or so but only once a day.

**Comment by dsarkar** [comment]
@KST1810 Thank you for your suggestion. We created an internal ticket. Any news on this issue will be published here.

---
Corona-Warn-App Open Source Team

**Comment by pkreissel** [comment]
To completely prevent people from e.g. photographing and tweeting QR-Codes (see janboehm on twitter) they would have to be renewed _after every scan_. This is not trivial however because it would require a privacy preserving backend that keeps track of scanned codes. But expiring the code every few seconds could already mitigate the problem.

**Comment by Ein-Tim** [comment]
@pkreissel 

I saw the [tweets](https://twitter.com/janboehm/status/1379573411400286209?s=21) from Jan Böhmermann on Twitter (and also [commented](https://twitter.com/eintim2/status/1379581566389735431?s=21) below one). 

I think for CWA it should be enough if venue owners change the code once a day, because:

1. According to the [documented threads - targeting specific venues](https://github.com/corona-warn-app/cwa-documentation/blob/master/event_registration.md#targeting-specific-venues), "CWA only allows a certain number of check-ins per day."
1.1 So if somebody want's to target venues in a noticeable scale, he needs more than one device running CWA (and he needs a positive result on more than one device, see point 2)

2. To warn other people via the Corona-Warn-App, you need to register a positive test result. This is only possible thorough either a QR-Code linked to a positive test result, this can only be obtained if somebody really is tested positive, or via a TAN requested from the Verification hotline 
2.1 The TeleTAN IHMO is the way an attacker would go, it's too hard to somehow get "positive QR-Codes". 
2.2 If somebody obtains a TeleTAN without having a positive test result, he/she already now is able to send out false positive warnings. 

3. The CWA system can't be "DDOS"sed by faking many check-ins. 
3.1 The luca app only hands the contacts of a positive tested over to the health authority, they then alert the contacts. 
If there was an event, which has been "DDOS"sed by many fake check-ins, luca tells the health authority that there were, for example, 10.000 people with you in this little restaurant. 
The health authority now would have to identify who really was there or warn all or no contacts.
3.2 The CWA works different. Even if there were 10.000 users checked-in with you at the same time (because somebody faked check-ins), all these 10.000 users can be notified, since the system is fully automatic. 
Thus, also your real contacts would be notified from CWA in such a "fake check-in" scenario. 

---

For sure, the from you described way is even safer, but as you said that would need a backend which keeps track of the QR-Codes (which belong to which event). 
If the QR-Code would only change after it was scanned, I also see one disadvantage:

The check-in would no longer work without internet connection, because the app would need to tell the server that the QR-Code has to be changed because somebody checked-in with it.

**Comment by jucktnich** [comment]
What's about dynamcaly generated qr codes on e.g. TV-Screens, wich are only valid for f.e. 5min. See #442

**Comment by dsarkar** [comment]
@Ein-Tim 

> The check-in would no longer work without internet connection, because the app would need to tell the server that the QR-Code has to be changed because somebody checked-in with it.

Not  necessarily: One could imagine i.e. that in the morning (or day before) the host of an event create a set of QR codes and the transfers this the tablet/screen with the "display-app".

**Comment by Ein-Tim** [comment]
@dsarkar 

Yeah right, but @pkreissel talked about a automatic switch of the QR-Code after it was scanned. 
This would only be possible with internet connection, or am I missing something here?

**Comment by jucktnich** [comment]
The host needs to know, that the code was scanned, but this could be done without internet connection, f.e. like I proposed with scanning a qr code by the host from the clients device.

**Comment by dsarkar** [comment]
Actually, the current implementation on https://www.coronawarn.app/en/eventregistration/ comes somewhat close to this proposal, with the difference that it is not "auto"-switching, but the create button has to be pressed whenever desired.

Nice-to have additionally:
- "show QR code full screen"-button 
- and an "renew QR code / create QR code"-button next to the QR code

**Comment by jucktnich** [comment]
There's a big difference I think, cause in this implementation, if you change the code, there's no warning between the two codes.
